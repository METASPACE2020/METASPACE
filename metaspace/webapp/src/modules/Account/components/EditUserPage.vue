<template>
  <div class="main-content">
    <div class="user-edit-page">
      <transfer-datasets-dialog
        v-if="showTransferDatasetsDialog"
        :currentUserId="currentUser && currentUser.id"
        :groupName="invitingGroup && invitingGroup.name"
        :isInvited="true"
        @accept="handleAcceptTransferDatasets"
        @close="handleCloseTransferDatasetsDialog"
      />
      <el-dialog
        title="Delete account"
        custom-class="delete-account-dialog"
        :visible.sync="showDeleteAccountDialog"
        width="500px"
        :lock-scroll="false">
        <p>If you delete your account, you will lose access to all private datasets.
        </p>
        <p>Please select whether you would like to delete all your datasets or keep them within METASPACE accessible by the group members:
        </p>
        <!--Changed the Text below to the above version after discussion with Theo, plz consider if it fits-->
        <!--If you delete your account, you will lose access to any datasets, groups and-->
        <!--projects that have been explicitly shared with you-->
        <el-checkbox v-model="delDatasets" style="margin: 10px 20px">
          Delete datasets that I have submitted
        </el-checkbox>
        <div style="margin: 10px 0 20px; text-align: right;">
          <el-button title="Cancel" @click="closeDeleteAccountDialog">Cancel</el-button>
          <el-button
            type="danger"
            title="Delete account"
            @click="deleteAccount()"
            :loading="isUserDeletionLoading">
            Delete account</el-button>
        </div>
        <p>
          <b>Note:</b> if you choose not to delete the datasets now, you will still be able to have them
          deleted later by emailing the <a href="mailto:contact@metaspace2020.eu">METASPACE administrators</a>.
        </p>
      </el-dialog>
      <el-row id="edit-user-page">
        <el-row>
          <el-col :span="16">
            <h2>User details</h2>
          </el-col>
          <el-col :span="8">
            <el-button title="Save"
                       type="primary"
                       @click="updateUserDetails"
                       class="saveButton"
                       :loading="isUserDetailsLoading">
              Save
            </el-button>
          </el-col>
        </el-row>
        <el-row :gutter="20">
          <el-form :disabled="isUserDetailsLoading" :rules="rules" :model="model" ref="form">
            <div class="user-details" style="padding-left: 15px;">
              <el-col :span="12">
                <div class="fullname">
                  <el-form-item prop="name" label="Full name:">
                    <el-input v-model="model.name" />
                  </el-form-item>
                </div>
              </el-col>
              <el-col :span="12">
                <div class="email">
                  <el-form-item prop="email" label="E-mail:">
                    <el-input v-model="model.email" />
                  </el-form-item>
                </div>
              </el-col>
            </div>
          </el-form>
        </el-row>

        <div class="groups">
          <h2>Groups</h2>
          <el-table
            :data="groupsData"
            style="width: 100%;padding-left: 15px;">
            <el-table-column
              prop="name"
              label="Group"
              width="180">
            </el-table-column>
            <el-table-column
              prop="roleName"
              label="Role"
              width="280">
            </el-table-column>
            <el-table-column
              prop="numDatasets"
              label="Dataset contributed">
            </el-table-column>
            <el-table-column>
              <template slot-scope="scope">
                <el-button
                  v-if="scope.row.role === 'MEMBER'"
                  size="mini"
                  icon="el-icon-arrow-right"
                  @click="leaveGroup('leave', scope.row)">
                  Leave
                </el-button><el-button
                  v-if="scope.row.role === 'PRINCIPAL INVESTIGATOR'"
                  size="mini"
                  icon="el-icon-arrow-right"
                  disabled>
                  Leave
                </el-button>
                <el-button
                  v-if="scope.row.role === 'INVITED'"
                  size="mini"
                  type="success"
                  @click="acceptInvitation(scope.row)"
                  icon="el-icon-check">
                  Accept
                </el-button>
                <el-button
                  v-if="scope.row.role === 'INVITED'"
                  size="mini"
                  icon="el-icon-close"
                  @click="leaveGroup('leaveGroup', scope.row)">
                  Decline
                </el-button>
              </template>
            </el-table-column>
          </el-table>
          <p>Primary group:</p>
          <el-select v-model="primaryGroupId" placeholder="Select"
                     style="padding-left: 15px;">
            <el-option
              v-for="item in groupsData"
              :key="item.id"
              :label="item.name"
              :value="item.id">
            </el-option>
          </el-select>
        </div>
        <!--The section below will be introduced in vFuture-->
        <!--<div class="notifications" style="margin-top: 30px">-->
          <!--<h2>Notifications</h2>-->
          <!--<div class="notification-list" style="padding-left: 10px">-->
            <!--<el-row :gutter="0">-->
              <!--<el-col :span="12"><p>Send an email when:</p>-->
                <!--<el-checkbox-group-->
                  <!--class="notifications">-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox-->
                      <!--label="My dataset's processing has successfully finished"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="My dataset's processing failed"></el-checkbox>-->
                  <!--</div>-->
                <!--</el-checkbox-group>-->
                <!--<p>Show a notification when:</p>-->
                <!--<el-checkbox-group :indeterminate="isIndeterminate" v-model="checkList"-->
                                   <!--class="notifications">-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="My dataset's processing has successfully finished"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="A dataset is added to a group or project that I belong to"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="A public dataset has been added to the queue"></el-checkbox>-->
                  <!--</div>-->
                  <!--<div class="notifications_checkbox">-->
                    <!--<el-checkbox class="notifications_checkbox"-->
                                 <!--label="A public dataset has finished processing"></el-checkbox>-->
                  <!--</div>-->
                <!--</el-checkbox-group>-->
              <!--</el-col>-->
            <!--</el-row>-->
          <!--</div>-->
        <!--</div>-->
        <div class="delete-account" style="margin-top: 45px">
          <h2>Delete account</h2>
          <p style="width: 100%;padding-left: 15px;">
            If you delete your METASPACE account, you can either delete all your datasets or keep them within METASPACE.
            For the latter, the private data will still be accessible by the group members only.
          </p>
        </div>
        <el-row>
          <el-button
            type="danger"
            title="Delete account"
            @click="openDeleteAccountDialog()"
            style="float:right; margin-top:15px">
            Delete account
          </el-button>
        </el-row>
      </el-row>
    </div>
  </div>
</template>

<script lang="ts">
  import Vue from 'vue'
  import { Component, Watch } from 'vue-property-decorator'
  import { updateUserMutation, deleteUserMutation, userProfileQuery, UserProfileQuery } from '../../../api/user';
  import { leaveGroupMutation, acceptGroupInvitationMutation } from '../../../api/group';
  import reportError from "../../../lib/reportError";
  import {refreshLoginStatus} from '../../../graphqlClient';
  import {ElForm} from "element-ui/types/form";
  import TransferDatasetsDialog from '../../GroupProfile/TransferDatasetsDialog.vue'
  import emailRegex from '../../../lib/emailRegex';
  import ConfirmAsync from '../../../components/ConfirmAsync';
  import { getRoleName, UserGroupRole } from '../../../api/group';

  interface Model {
    name: string;
    email: string | null;
  }

  interface GroupRow {
    id: string;
    name: string;
    role: UserGroupRole;
    roleName: string;
    numDatasets: number;
  }

  @Component({
    components: {
      TransferDatasetsDialog
    },
    apollo: {
      currentUser: {
        query: userProfileQuery
      }
    }
  })

  export default class EditUserPage extends Vue {
    showDeleteAccountDialog: boolean = false;
    isUserDetailsLoading: boolean = false;
    isUserDeletionLoading: boolean = false;
    showTransferDatasetsDialog: boolean = false;

    currentUser: UserProfileQuery | null = null;
    model: Model = {
      name: '',
      email: ''
    };
    primaryGroupId: string | null = null;
    invitingGroup: GroupRow | null = null;

    delDatasets: boolean = false;
    rules: object = {
      name: [
        {required: true, min: 3, max: 50 , message: 'Please enter a correct fullname', trigger: "blur"}
      ],
      email: [
        {required: true, pattern: emailRegex,
        message: 'Please enter a correct Email address', trigger: "blur"
      }]
    };

    @Watch('currentUser', {deep: true})
    onCurrentUserChanged(this: any) {
      this.model.name = this.currentUser.name;
      this.model.email = this.currentUser.email;
      this.primaryGroupId = this.currentUser.primaryGroup ? this.currentUser.primaryGroup.group.id : null;
    }

    openDeleteAccountDialog() {
      this.showDeleteAccountDialog = true;
    }

    closeDeleteAccountDialog() {
      this.showDeleteAccountDialog = false;
    }

    get groupsData(): GroupRow[] {
      if (this.currentUser == null || this.currentUser.groups == null) {
        return [];
      }
      return this.currentUser.groups.map(({group, numDatasets, role}) => {
        const {id, name} = group;
        const roleName = getRoleName(role);
        return {id, name, role, roleName, numDatasets}
      });
    }

    async updateUserDetails() {
      try {
        await (this.$refs.form as ElForm).validate();
      } catch (err) {
        return;
      }
      try {
        if (this.currentUser!.email !== this.model.email) {
          try {
            await this.$confirm(
              "Are you sure you want to change email address? A verification email will be sent to your new address to confirm the change.",
              "Confirm email address change", {
                confirmButtonText: "Yes, send verification email",
                lockScroll: false
              });
          } catch {
            return
          }
        }
        this.isUserDetailsLoading = true;
        await this.$apollo.mutate({
          mutation: updateUserMutation,
          variables: {
            update: {
              id: this.currentUser!.id,
              name: this.model.name,
              email: this.model.email,
              primaryGroupId: this.primaryGroupId
            }
          },
        });
        await this.$apollo.queries.currentUser.refetch();
          this.$message({
            type: "success",
            message: "A verification link has been sent to your new email address. " +
            "Please click the link in this email to confirm the change."
          });
      } catch(err) {
        reportError(err);
      } finally {
        this.isUserDetailsLoading = false;
      }
    }

    @ConfirmAsync(function(action: string, groupRow: GroupRow) {
      return {
        message: (action === 'leave') ? `Are you sure you want to leave ${groupRow.name}?` :
          `Are you sure you want to decline the invitation to ${groupRow.name}?`,
        confirmButtonText: (action === 'leave') ? "Yes, leave the group" :
          "Yes, decline the invitation",
        confirmButtonLoadingText: 'Leaving...'
      };
    })
    async leaveGroup(action: string, groupRow: GroupRow) {
      await this.$apollo.mutate({
        mutation: leaveGroupMutation,
        variables: {
          groupId: groupRow.id
        }
      });
      this.$message({
        message: "You have declined the invitation"
      });
      await this.$apollo.queries.currentUser.refetch();
      if (action === 'leave') {
        this.$message({
          message: "You have successfully left the group"
        });
      }
    }

    async deleteAccount(this: any) {
      try {
        this.isUserDeletionLoading = true;
        await this.$apollo.mutate({
          mutation: deleteUserMutation,
          variables: {
            id: this.currentUser.id,
            deleteDatasets: this.delDatasets
          }
        });
        this.$message({
          type: "success",
          message: "You have successfully deleted your account!"
        })
      } catch(err) {
        reportError(err);
      } finally {
        await refreshLoginStatus();
        this.closeDeleteAccountDialog();
        this.isUserDeletionLoading = false;
      }
    }

    async acceptInvitation(groupRow: GroupRow) {
      this.showTransferDatasetsDialog = true;
      this.invitingGroup = groupRow;
    }

    async handleAcceptTransferDatasets(selectedDatasetIds: string[]) {
      try {
        await this.$apollo.mutate({
          mutation: acceptGroupInvitationMutation,
          variables: { groupId: this.invitingGroup!.id, bringDatasets: selectedDatasetIds },
        });
        await this.$apollo.queries.currentUser.refetch();
        this.$message({
          type: "success",
          message: "You have successfully joined the group!"
        });
        this.showTransferDatasetsDialog = false;
      } catch(err) {
        reportError(err);
      } finally {
        this.showTransferDatasetsDialog = false;
      }
    }

    handleCloseTransferDatasetsDialog() {
      this.showTransferDatasetsDialog = false;
    }
  }
</script>

<style scoped>
  .main-content {
    padding: 100px 20px 20px 20px;
    display: flex;
    justify-content: center;
  }

  .user-edit-page {
    max-width: 950px;
  }

  .saveButton {
    width: 100px;
    padding: 8px;
    float: right;
    margin: 20px 0;
  }

  .user-details {
    display: inline-block;
    width: 600px;
  }

  /deep/ .delete-account-dialog .el-dialog__body {
    padding: 0 20px 20px 20px;
  }

  /* Uncomment when the vFuture notifications will be introduced
  /*.notifications_checkbox {*/
    /*margin-left: 0;*/
    /*padding: 0;*/
  /*}*/
</style>
