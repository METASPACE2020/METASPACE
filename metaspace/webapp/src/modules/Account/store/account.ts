import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
