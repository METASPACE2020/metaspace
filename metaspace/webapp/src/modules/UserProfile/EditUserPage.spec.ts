import { nextTick, ref } from 'vue'
import { flushPromises, mount } from '@vue/test-utils'
import EditUserPage from './EditUserPage.vue'
import { initMockGraphqlClient } from '../../tests/utils/mockGraphqlClient'
import { vi } from 'vitest'
import { DefaultApolloClient, useMutation, useQuery } from '@vue/apollo-composable'
import store from '../../store'
import router from '../../router'
import { ElMessageBox } from '../../lib/element-plus'

vi.mock('element-plus', async () => {
  const actual: any = await vi.importActual('element-plus')
  return {
    ...actual,
    ElMessageBox: {
      confirm: vi.fn(),
    },
  }
})

vi.mock('@vue/apollo-composable', () => ({
  useQuery: vi.fn(),
  useMutation: vi.fn(),
  useSubscription: vi.fn(() => ({ onResult: vi.fn() })),
  DefaultApolloClient: vi.fn(),
}))

let graphqlMocks: any
describe('EditUserPage', () => {
  const mockCurrentUser = {
    id: '22333',
    name: 'foo',
    email: 'foo@bar.baz',
    role: 'user',
    groups: [
      { role: 'MEMBER', numDatasets: 0, group: { id: 'AAA', name: 'Group A', urlSlug: 'grp-a' } },
      { role: 'INVITED', numDatasets: 0, group: { id: 'BBB', name: 'Group B', urlSlug: null } },
      { role: 'PENDING', numDatasets: 0, group: { id: 'CCC', name: 'Group C', urlSlug: null } },
      { role: 'GROUP_ADMIN', numDatasets: 20, group: { id: 'DDD', name: 'Group D', urlSlug: null } },
    ],
    primaryGroup: { role: 'GROUP_ADMIN', numDatasets: 20, group: { id: 'DDD', name: 'Group D', urlSlug: null } },
    projects: [
      { role: 'MEMBER', numDatasets: 0, project: { id: 'AA', name: 'Project A', urlSlug: 'proj-a' } },
      { role: 'INVITED', numDatasets: 0, project: { id: 'BB', name: 'Project B', urlSlug: null } },
      { role: 'PENDING', numDatasets: 0, project: { id: 'CC', name: 'Project C', urlSlug: null } },
      { role: 'MANAGER', numDatasets: 20, project: { id: 'DD', name: 'Project D', urlSlug: null } },
    ],
  }
  const mockUpdateUserMutation = vi.fn(() => ({}))
  const graphqlMock = async () => {
    const queryParams = {
      currentUser: () => mockCurrentUser,
    }
    const mutateParams = {
      updateUser: mockUpdateUserMutation,
    }
    graphqlMocks = await initMockGraphqlClient({
      Query: () => queryParams,
      Mutation: () => mutateParams,
    })
    ;(useQuery as any).mockReturnValue({
      result: ref(Object.keys(queryParams).reduce((acc, key) => ({ ...acc, [key]: queryParams[key]() }), {})),
      loading: ref(false),
      onResult: vi.fn(),
      refetch: vi.fn(),
    })
    ;(useMutation as any).mockReturnValue({
      mutate: mockUpdateUserMutation,
    })
  }

  beforeEach(async () => {
    vi.clearAllMocks()
    await graphqlMock()
  })

  it('should match snapshot', async () => {
    const wrapper = mount(EditUserPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to submit changes to the user', async () => {
    const wrapper = mount(EditUserPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    const saveButton = wrapper.find('.saveButton')
    const name = 'foo bar'
    const email = 'changed@bar.baz'

    wrapper.vm.model.name = name
    wrapper.vm.model.email = email
    saveButton.trigger('click')

    await flushPromises()
    await nextTick()

    expect(ElMessageBox.confirm).toHaveBeenCalledWith(
      'Are you sure you want to change email address? ' +
        'A verification email will be sent to your new address to confirm the change.',
      'Confirm email address change',
      {
        confirmButtonText: 'Yes, send verification email',
        lockScroll: false,
      }
    )

    expect(mockUpdateUserMutation).toHaveBeenCalledTimes(1)
  })

  it('should not include unchanged fields in the update payload', async () => {
    const wrapper = mount(EditUserPage, {
      global: {
        plugins: [store, router],
        provide: {
          [DefaultApolloClient]: graphqlMocks,
        },
      },
    })
    await flushPromises()
    await nextTick()

    const saveButton = wrapper.find('.saveButton')
    const name = 'foo bar'

    wrapper.vm.model.name = name
    saveButton.trigger('click')

    await flushPromises()
    await nextTick()

    expect(ElMessageBox.confirm).not.toHaveBeenCalledWith(
      'Are you sure you want to change email address? ' +
        'A verification email will be sent to your new address to confirm the change.',
      'Confirm email address change',
      {
        confirmButtonText: 'Yes, send verification email',
        lockScroll: false,
      }
    )

    expect(mockUpdateUserMutation).toHaveBeenCalledTimes(1)
  })
})
