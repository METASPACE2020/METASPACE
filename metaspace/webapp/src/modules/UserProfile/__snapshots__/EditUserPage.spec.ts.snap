// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserPage should match snapshot 1`] = `
<div class="main-content">
  <mock-el-dialog title="Delete account" custom-class="delete-account-dialog" width="500px">
    <p>If you delete your account, you will lose access to all private datasets.
    </p>
    <p>Please select whether you would like to delete all your datasets or keep them within METASPACE accessible by the group members:
    </p>
    <label role="checkbox" class="el-checkbox" style="margin: 10px 20px;"><span aria-checked="mixed" class="el-checkbox__input"><span class="el-checkbox__inner"></span>
      <input type="checkbox" aria-hidden="true" class="el-checkbox__original" value="">
      </span><span class="el-checkbox__label">
      Delete datasets that I have submitted
    <!----></span></label>
    <div style="margin: 10px 0px 20px; text-align: right;">
      <button type="button" class="el-button el-button--default" title="Cancel">
        <!---->
        <!----><span>Cancel</span></button>
      <button type="button" class="el-button el-button--danger" title="Delete account">
        <!---->
        <!----><span>
        Delete account</span></button>
    </div>
    <p>
      <b>Note:</b> if you choose not to delete the datasets now, you will still be able to have them deleted later by emailing the
      <a href="mailto:contact@metaspace2020.eu">METASPACE administrators</a>.
    </p>
  </mock-el-dialog>
  <div class="user-edit-page" mock-v-loading="false">
    <div class="el-row">
      <div class="el-col el-col-16">
        <h2>User details</h2>
      </div>
      <div class="el-col el-col-8">
        <button type="button" class="el-button saveButton el-button--primary" title="Save">
          <!---->
          <!----><span>
          Save
        </span></button>
      </div>
    </div>
    <div class="el-row" style="margin-left: -10px; margin-right: -10px;">
      <form class="el-form">
        <div class="user-details" style="padding-left: 15px;">
          <div class="el-col el-col-12" style="padding-left: 10px; padding-right: 10px;">
            <div>
              <div class="el-form-item is-required">
                <label for="name" class="el-form-item__label">Full name</label>
                <div class="el-form-item__content">
                  <div class="el-input">
                    <!---->
                    <input type="text" autocomplete="off" name="name" class="el-input__inner">
                    <!---->
                    <!---->
                    <!---->
                  </div>
                  <mock-transition name="el-zoom-in-top"></mock-transition>
                </div>
              </div>
            </div>
          </div>
          <div class="el-col el-col-12" style="padding-left: 10px; padding-right: 10px;">
            <div>
              <div class="el-form-item is-required">
                <label for="email" class="el-form-item__label">Email address</label>
                <div class="el-form-item__content">
                  <div class="el-input">
                    <!---->
                    <input type="text" autocomplete="off" name="email" class="el-input__inner">
                    <!---->
                    <!---->
                    <!---->
                  </div>
                  <mock-transition name="el-zoom-in-top"></mock-transition>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div>
      <h2>Groups</h2>
      <div>
        <!---->
        <div class="el-table el-table--fit el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition" style="width: 100%; padding-left: 15px;">
          <div class="hidden-columns">
            <div></div>
            <div></div>
            <div></div>
            <div></div>
          </div>
          <div class="el-table__header-wrapper">
            <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 620px;">
              <colgroup>
                <col name="el-table_1_column_1" width="180">
                <col name="el-table_1_column_2" width="280">
                <col name="el-table_1_column_3" width="80">
                <col name="el-table_1_column_4" width="80">
              </colgroup>
              <thead class="">
                <tr class="">
                  <th colspan="1" rowspan="1" class="el-table_1_column_1     is-leaf">
                    <div class="cell">Group</div>
                  </th>
                  <th colspan="1" rowspan="1" class="el-table_1_column_2     is-leaf">
                    <div class="cell">Role</div>
                  </th>
                  <th colspan="1" rowspan="1" class="el-table_1_column_3     is-leaf">
                    <div class="cell">Datasets contributed</div>
                  </th>
                  <th colspan="1" rowspan="1" class="el-table_1_column_4     is-leaf">
                    <div class="cell"></div>
                  </th>
                </tr>
              </thead>
            </table>
          </div>
          <div class="el-table__body-wrapper is-scrolling-left">
            <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 620px;">
              <colgroup>
                <col name="el-table_1_column_1" width="180">
                <col name="el-table_1_column_2" width="280">
                <col name="el-table_1_column_3" width="80">
                <col name="el-table_1_column_4" width="80">
              </colgroup>
              <tbody>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                    <div class="cell">
                      <a href="#/group/AAA" class="">Group A</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                    <div class="cell">Member</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                    <div class="cell">
                      <!----><span>0</span></div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                    <div class="cell">
                      <button type="button" class="el-button el-button--default el-button--mini">
                        <!---->
                        <i class="el-icon-arrow-right"></i><span>
          Leave
        </span></button>
                      <!---->
                      <!---->
                      <!---->
                    </div>
                  </td>
                </tr>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                    <div class="cell">
                      <a href="#/group/BBB" class="">Group B</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                    <div class="cell">Invited</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                    <div class="cell">
                      <!----><span>0</span></div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                    <div class="cell">
                      <!---->
                      <!---->
                      <button type="button" class="el-button el-button--success el-button--mini">
                        <!---->
                        <i class="el-icon-check"></i><span>
          Accept
        </span></button>
                      <button type="button" class="el-button el-button--default el-button--mini">
                        <!---->
                        <i class="el-icon-close"></i><span>
          Decline
        </span></button>
                    </div>
                  </td>
                </tr>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                    <div class="cell">
                      <a href="#/group/CCC" class="">Group C</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                    <div class="cell">Requesting access</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                    <div class="cell">
                      <!----><span>0</span></div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                    <div class="cell">
                      <!---->
                      <!---->
                      <!---->
                      <!---->
                    </div>
                  </td>
                </tr>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                    <div class="cell">
                      <a href="#/group/DDD" class="">Group D</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                    <div class="cell">Principal investigator</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                    <div class="cell">
                      <a href="#/datasets?grp=DDD&amp;subm=22333" class="">
                        20
                      </a>
                      <!---->
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                    <div class="cell">
                      <!---->
                      <button disabled="disabled" type="button" class="el-button el-button--default el-button--mini is-disabled">
                        <!---->
                        <i class="el-icon-arrow-right"></i><span>
          Leave
        </span></button>
                      <!---->
                      <!---->
                    </div>
                  </td>
                </tr>
                <!---->
              </tbody>
            </table>
            <!---->
            <!---->
          </div>
          <!---->
          <!---->
          <!---->
          <!---->
          <div class="el-table__column-resize-proxy" style="display: none;"></div>
        </div>
      </div>
      <div>
        <p>Primary group:</p>
        <mock-el-select placeholder="Select" style="padding-left: 15px;">
          <mock-el-option label="Group A" value="AAA"></mock-el-option>
          <mock-el-option label="Group B" value="BBB"></mock-el-option>
          <mock-el-option label="Group C" value="CCC"></mock-el-option>
          <mock-el-option label="Group D" value="DDD"></mock-el-option>
        </mock-el-select>
      </div>
    </div>
    <div style="margin-top: 40px;">
      <h2>Projects</h2>
      <div>
        <div class="el-table el-table--fit el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition" style="width: 100%; padding-left: 15px;">
          <div class="hidden-columns">
            <div></div>
            <div></div>
            <div></div>
            <div></div>
          </div>
          <div class="el-table__header-wrapper">
            <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 620px;">
              <colgroup>
                <col name="el-table_2_column_5" width="180">
                <col name="el-table_2_column_6" width="280">
                <col name="el-table_2_column_7" width="80">
                <col name="el-table_2_column_8" width="80">
              </colgroup>
              <thead class="">
                <tr class="">
                  <th colspan="1" rowspan="1" class="el-table_2_column_5     is-leaf">
                    <div class="cell">Project</div>
                  </th>
                  <th colspan="1" rowspan="1" class="el-table_2_column_6     is-leaf">
                    <div class="cell">Role</div>
                  </th>
                  <th colspan="1" rowspan="1" class="el-table_2_column_7     is-leaf">
                    <div class="cell">Datasets contributed</div>
                  </th>
                  <th colspan="1" rowspan="1" class="el-table_2_column_8     is-leaf">
                    <div class="cell"></div>
                  </th>
                </tr>
              </thead>
            </table>
          </div>
          <div class="el-table__body-wrapper is-scrolling-left">
            <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 620px;">
              <colgroup>
                <col name="el-table_2_column_5" width="180">
                <col name="el-table_2_column_6" width="280">
                <col name="el-table_2_column_7" width="80">
                <col name="el-table_2_column_8" width="80">
              </colgroup>
              <tbody>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_2_column_5  ">
                    <div class="cell">
                      <a href="#/project/AA" class="">Project A</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_6  ">
                    <div class="cell">Member</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_7  ">
                    <div class="cell">
                      <!----><span>0</span></div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_8  ">
                    <div class="cell">
                      <button type="button" class="el-button el-button--default el-button--mini">
                        <!---->
                        <i class="el-icon-arrow-right"></i><span>
          Leave
        </span></button>
                      <!---->
                      <!---->
                      <!---->
                    </div>
                  </td>
                </tr>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_2_column_5  ">
                    <div class="cell">
                      <a href="#/project/BB" class="">Project B</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_6  ">
                    <div class="cell">Invited</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_7  ">
                    <div class="cell">
                      <!----><span>0</span></div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_8  ">
                    <div class="cell">
                      <!---->
                      <!---->
                      <button type="button" class="el-button el-button--success el-button--mini">
                        <!---->
                        <i class="el-icon-check"></i><span>
          Accept
        </span></button>
                      <button type="button" class="el-button el-button--default el-button--mini">
                        <!---->
                        <i class="el-icon-close"></i><span>
          Decline
        </span></button>
                    </div>
                  </td>
                </tr>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_2_column_5  ">
                    <div class="cell">
                      <a href="#/project/CC" class="">Project C</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_6  ">
                    <div class="cell">Requesting access</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_7  ">
                    <div class="cell">
                      <!----><span>0</span></div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_8  ">
                    <div class="cell">
                      <!---->
                      <!---->
                      <!---->
                      <!---->
                    </div>
                  </td>
                </tr>
                <tr class="el-table__row">
                  <td rowspan="1" colspan="1" class="el-table_2_column_5  ">
                    <div class="cell">
                      <a href="#/project/DD" class="">Project D</a>
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_6  ">
                    <div class="cell">Project manager</div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_7  ">
                    <div class="cell">
                      <a href="#/datasets?prj=DD&amp;subm=22333" class="">
                        20
                      </a>
                      <!---->
                    </div>
                  </td>
                  <td rowspan="1" colspan="1" class="el-table_2_column_8  ">
                    <div class="cell">
                      <!---->
                      <button disabled="disabled" type="button" class="el-button el-button--default el-button--mini is-disabled">
                        <!---->
                        <i class="el-icon-arrow-right"></i><span>
          Leave
        </span></button>
                      <!---->
                      <!---->
                    </div>
                  </td>
                </tr>
                <!---->
              </tbody>
            </table>
            <!---->
            <!---->
          </div>
          <!---->
          <!---->
          <!---->
          <!---->
          <div class="el-table__column-resize-proxy" style="display: none;"></div>
        </div>
      </div>
    </div>
    <div style="margin-top: 45px;">
      <h2>Delete account</h2>
      <p style="width: 100%; padding-left: 15px;">
        If you delete your METASPACE account, you can either delete all your datasets or keep them within METASPACE. For the latter, the private data will still be accessible by the group members only.
      </p>
    </div>
    <div class="el-row">
      <button type="button" class="el-button el-button--danger" title="Delete account" style="float: right; margin-top: 15px;">
        <!---->
        <!----><span>
        Delete account
      </span></button>
    </div>
  </div>
</div>
`;
